test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/DerivedTypes.cs

```diff
@@ -0,0 +1,360 @@
+        }
+
+        public string Protect(AuthenticationProperties data)
+        {
+            return _protect;
+        }
+
+        AuthenticationProperties ISecureDataFormat<AuthenticationProperties>.Unprotect(string protectedText)
+        {
+            return _unprotect;
+        }
+    }
+
+    public class CustomConfigureOptions : ConfigureOptions<OpenIdConnectAuthenticationOptions>
+    {
+        public CustomConfigureOptions(Action<OpenIdConnectAuthenticationOptions> action)
+            : base(action)
+        {
+        }
+
+        public override void Configure(OpenIdConnectAuthenticationOptions options, string name = "")
+        {
+            base.Configure(options, name);
+            return;
+        }
+    }
+
+    public class CustomLogger : ILogger, IDisposable
+    {
+        LogLevel _logLevel = 0;
+
+        public CustomLogger(LogLevel logLevel = LogLevel.Debug)
+        {
+            _logLevel = logLevel;
+        }
+
+        List<LogEntry> _logEntries = new List<LogEntry>();
+
+        public IDisposable BeginScopeImpl(object state)
+        {
+            return this;
+        }
+
+        public void Dispose()
+        {
+        }
+
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return (logLevel >= _logLevel);
+        }
+
+        public void Log(LogLevel logLevel, int eventId, object state, Exception exception, Func<object, Exception, string> formatter)
+        {
+            if (IsEnabled(logLevel))
+            {                
+                var logEntry =
+                    new LogEntry
+                    {
+                        EventId = eventId,
+                        Exception = exception,
+                        Formatter = formatter,
+                        Level = logLevel,
+                        State = state,
+                    };
+
+                _logEntries.Add(logEntry);
+                Debug.WriteLine(logEntry.ToString());
+            }
+        }
+
+        public List<LogEntry> Logs { get { return _logEntries; } }
+    }
+
+    public class CustomLoggerFactory : ILoggerFactory
+    {
+        CustomLogger _logger;
+        LogLevel _logLevel = LogLevel.Debug;
+
+        public CustomLoggerFactory(LogLevel logLevel)
+        {
+            _logLevel = logLevel;
+            _logger = new CustomLogger(_logLevel);
+        }
+
+        public LogLevel MinimumLevel
+        {
+            get { return _logLevel; }
+            set { _logLevel = value; }
+        }
+
+        public void AddProvider(ILoggerProvider provider)
+        {
+        }
+
+        public ILogger CreateLogger(string categoryName)
+        {
+            return _logger;
+        }
+
+        public CustomLogger Logger { get { return _logger; } }
+    }
+
+    /// <summary>
+    /// Allows for controlling methods and capturing and exposing artifacts.
+    /// </summary>
+    public class CustomOpenIdConnectAuthenticationHandler : OpenIdConnectAuthenticationHandler
```

Comment:
Should look into using Moq for these types of mocks. It makes the actual test code a lot more readable because they end up being largely self-contained. Using extra types can make it more difficult to understand a test because it's not always clear what's being customized vs. what's standard behavior.